src/device/sound.rs

```diff
@@ -0,0 +1,1563 @@
+        // rsp is just a header, so it can be compared with VirtIOSndHdr
+        if rsp == VirtIOSndHdr::from(RequestStatusCode::VirtioSndSOk) {
+            Ok(())
+        } else {
+            Err(Error::IoError)
+        }
+    }
+
+    /// Start a stream with specified stream ID.
+    pub fn pcm_start(&mut self, stream_id: u32) -> Result {
+        if !self.set_up {
+            self.set_up();
+            self.set_up = true;
+        }
+        let request_hdr = VirtIOSndHdr::from(CommandCode::VirtioSndRPcmStart);
+        let rsp: VirtIOSndHdr = self.request(VirtIOSndPcmHdr {
+            hdr: request_hdr,
+            stream_id,
+        })?;
+        // rsp is just a header, so it can be compared with VirtIOSndHdr
+        if rsp == VirtIOSndHdr::from(RequestStatusCode::VirtioSndSOk) {
+            Ok(())
+        } else {
+            Err(Error::IoError)
+        }
+    }
+
+    /// Stop a stream with specified stream ID.
+    pub fn pcm_stop(&mut self, stream_id: u32) -> Result {
+        if !self.set_up {
+            self.set_up();
+            self.set_up = true;
+        }
+        let request_hdr = VirtIOSndHdr::from(CommandCode::VirtioSndRPcmStop);
+        let rsp: VirtIOSndHdr = self.request(VirtIOSndPcmHdr {
+            hdr: request_hdr,
+            stream_id,
+        })?;
+        // rsp is just a header, so it can be compared with VirtIOSndHdr
+        if rsp == VirtIOSndHdr::from(RequestStatusCode::VirtioSndSOk) {
+            Ok(())
+        } else {
+            Err(Error::IoError)
+        }
+    }
+
+    /// Transfer PCM frame to device, based on the stream type(OUTPUT/INPUT).
+    /// 
+    /// This is a blocking method that will not return until the audio playback is complete.
+    pub fn pcm_xfer(&mut self, stream_id: u32, frames: &[u8]) -> Result {
+        const U32_SIZE: usize = mem::size_of::<u32>();
+        if !self.set_up {
+            self.set_up();
+            self.set_up = true;
+        }
+        if !self.pcm_parameters[stream_id as usize].setup {
+            warn!("Please set parameters for a stream before using it!");
+            return Err(Error::IoError);
+        }
+        let buffer_size = self.pcm_parameters[stream_id as usize].buffer_bytes as usize;
+        assert!(buffer_size * 2 <= frames.len());
+        let mut buf1 = vec![0; U32_SIZE + buffer_size];
+        let mut buf2 = vec![0; U32_SIZE + buffer_size];
+
+        buf1[..U32_SIZE].copy_from_slice(&stream_id.to_le_bytes());
+        buf1[U32_SIZE..U32_SIZE + buffer_size].copy_from_slice(&frames[..buffer_size]);
+
+        buf2[..U32_SIZE].copy_from_slice(&stream_id.to_le_bytes());
+        buf2[U32_SIZE..U32_SIZE + buffer_size]
+            .copy_from_slice(&frames[buffer_size..buffer_size * 2]);
+
+        let mut outputs = vec![0; 32];
+        let mut token1 = unsafe { self.tx_queue.add(&[&buf1], &mut [&mut outputs]).unwrap() };
+        let mut token2 = unsafe { self.tx_queue.add(&[&buf2], &mut [&mut outputs]).unwrap() };
```

Comment:
Adding two entries to the queue with the same output buffer is unsound. There must not be any other reference to an output buffer until it is popped.

Reply:
I'm not going to do anything with this `outputs` here; it merely serves the purpose of receiving return values sent by the device.

Reply:
I've rewritten this to fix the issue.